sound/soc/intel/boards/sof_maxim_common.c

```diff
@@ -10,67 +10,80 @@
+        unsigned int tx;
+        unsigned int rx;
+} maxim_tdm_mask[] = {
+        {.tx = 0x3, .rx = 0x3},
+        {.tx = 0xC, .rx = 0x3},
+};
 
-const struct snd_soc_dapm_route max_98373_dapm_routes[] = {
-	/* speaker */
+static const struct {
+        unsigned int tx;
+        unsigned int rx;
+} maxim_dsp_tdm_mask[] = {
+        {.tx = 0x1, .rx = 0x3},
+        {.tx = 0x2, .rx = 0x3},
+        {.tx = 0x4, .rx = 0x3},
+        {.tx = 0x8, .rx = 0x3},
+};
+
+static const struct snd_kcontrol_new maxim_kcontrols[] = {
+	SOC_DAPM_PIN_SWITCH("Left Spk"),
+	SOC_DAPM_PIN_SWITCH("Right Spk"),
+};
+
+static const struct snd_soc_dapm_widget maxim_dapm_widgets[] = {
+	SND_SOC_DAPM_SPK("Left Spk", NULL),
+	SND_SOC_DAPM_SPK("Right Spk", NULL),
+};
+
+static const struct snd_soc_dapm_route maxim_dapm_routes[] = {
 	{ "Left Spk", NULL, "Left BE_OUT" },
 	{ "Right Spk", NULL, "Right BE_OUT" },
 };
-EXPORT_SYMBOL_NS(max_98373_dapm_routes, SND_SOC_INTEL_SOF_MAXIM_COMMON);
 
-static struct snd_soc_codec_conf max_98373_codec_conf[] = {
-	{
-		.dlc = COMP_CODEC_CONF(MAX_98373_DEV0_NAME),
-		.name_prefix = "Right",
-	},
-	{
-		.dlc = COMP_CODEC_CONF(MAX_98373_DEV1_NAME),
-		.name_prefix = "Left",
-	},
+static const struct snd_kcontrol_new maxim_tt_kcontrols[] = {
+	SOC_DAPM_PIN_SWITCH("TL Spk"),
+	SOC_DAPM_PIN_SWITCH("TR Spk"),
 };
 
-struct snd_soc_dai_link_component max_98373_components[] = {
-	{  /* For Right */
-		.name = MAX_98373_DEV0_NAME,
-		.dai_name = MAX_98373_CODEC_DAI,
-	},
-	{  /* For Left */
-		.name = MAX_98373_DEV1_NAME,
-		.dai_name = MAX_98373_CODEC_DAI,
-	},
+static const struct snd_soc_dapm_widget maxim_tt_dapm_widgets[] = {
+	SND_SOC_DAPM_SPK("TL Spk", NULL),
+	SND_SOC_DAPM_SPK("TR Spk", NULL),
 };
-EXPORT_SYMBOL_NS(max_98373_components, SND_SOC_INTEL_SOF_MAXIM_COMMON);
 
-static int max_98373_hw_params(struct snd_pcm_substream *substream,
-			       struct snd_pcm_hw_params *params)
-{
-	struct snd_soc_pcm_runtime *rtd = asoc_substream_to_rtd(substream);
-	struct snd_soc_dai *codec_dai;
-	int j;
-
-	for_each_rtd_codec_dais(rtd, j, codec_dai) {
-		if (!strcmp(codec_dai->component->name, MAX_98373_DEV0_NAME)) {
-			/* DEV0 tdm slot configuration */
-			snd_soc_dai_set_tdm_slot(codec_dai, 0x03, 3, 8, 32);
-		}
-		if (!strcmp(codec_dai->component->name, MAX_98373_DEV1_NAME)) {
-			/* DEV1 tdm slot configuration */
-			snd_soc_dai_set_tdm_slot(codec_dai, 0x0C, 3, 8, 32);
-		}
-	}
-	return 0;
-}
+static const struct snd_soc_dapm_route maxim_tt_dapm_routes[] = {
+	/* Tweeter speaker */
+	{ "TL Spk", NULL, "Tweeter Left BE_OUT" },
+	{ "TR Spk", NULL, "Tweeter Right BE_OUT" },
+};
 
-int max_98373_trigger(struct snd_pcm_substream *substream, int cmd)
+int maxim_trigger(struct snd_pcm_substream *substream, int cmd)
 {
 	struct snd_soc_pcm_runtime *rtd = asoc_substream_to_rtd(substream);
 	struct snd_soc_dai *codec_dai;
 	struct snd_soc_dai *cpu_dai;
 	int j;
 	int ret = 0;
 
+	if (strcmp(maxim_components[0].dai_name, "max98373-aif1"))
+		return 0;
```

Comment:
this seems to NOT be generic code, so what was the point of merging the functions if now we have codec-specific stuff in the helpers?

Reply:
@plbossart , my beginning thoughts was reducing the .ops codes. For better maintenance, I should add codec_ops separately, any advices?

const struct snd_soc_ops max_98373_ops = {
	.hw_params = max98373_hw_params,
	.trigger = 98373_trigger,
};
const struct snd_soc_ops max98390_ops = {
	.hw_params = max98390_hw_params,
};
const struct snd_soc_ops max_98396_ops = {
	.hw_params = max98373_hw_params,
};